File: shared/src/main/scala/stasis/shared/ops/Operation.scala
package stasis.shared.ops

import akka.Done

import java.time.Instant
import scala.concurrent.Future

trait Operation {
  def id: Operation.Id
  def start(): Future[Done]
  def stop(): Unit
  def `type`: Operation.Type
}

object Operation {
  type Id = java.util.UUID

  def generateId(): Id = java.util.UUID.randomUUID()

  sealed trait Type
  object Type {
    sealed trait ClientBased extends Type
    sealed trait ServerBased extends Type

    case object Backup extends ClientBased
    case object Recovery extends ClientBased
    case object Expiration extends ClientBased
    case object Validation extends ClientBased
    case object KeyRotation extends ClientBased

    case object GarbageCollection extends ServerBased
  }

  final case class Progress(
    started: Instant,
    total: Int,
    processed: Int,
    failures: Int,
    completed: Option[Instant]
  )
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/Operation.kt
package stasis.client_android.lib.ops

import kotlinx.coroutines.CoroutineScope
import java.time.Instant
import java.util.UUID

interface Operation {
    val id: OperationId
    fun start(withScope: CoroutineScope, f: (Throwable?) -> Unit)
    fun stop()
    fun type(): Type

    companion object {
        fun generateId(): OperationId = UUID.randomUUID()
    }

    sealed class Type {
        override fun toString(): String = javaClass.simpleName

        object Backup : Type()
        object Recovery : Type()
        object Expiration : Type()
        object Validation : Type()
        object KeyRotation : Type()
        object GarbageCollection : Type()

        companion object {
            fun fromString(string: String): Type =
                when (string) {
                    "Backup" -> Backup
                    "Recovery" -> Recovery
                    "Expiration" -> Expiration
                    "Validation" -> Validation
                    "KeyRotation" -> KeyRotation
                    "GarbageCollection" -> GarbageCollection
                    else -> throw IllegalArgumentException("Unexpected operation type provided: [$string]")
                }
        }
    }

    data class Progress(
        val started: Instant,
        val total: Int,
        val processed: Int,
        val failures: Int,
        val completed: Instant?
    )

    sealed class Restriction {
        data object NoConnection : Restriction()
        data object LimitedNetwork : Restriction()
    }
}

typealias OperationId = UUID
